cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/offer/PersistentOfferRequirementProvider.java

```diff
@@ -70,16 +72,42 @@ public OfferRequirement getNewOfferRequirement(CassandraContainer container) {
         }
     }
 
-    public OfferRequirement getReplacementOfferRequirement(Protos.TaskInfo taskInfo) {
-        LOGGER.info("Getting replacement requirement for task: {}", taskInfo.getTaskId().getValue());
-        return getExistingOfferRequirement(taskInfo);
+    public OfferRequirement getReplacementOfferRequirement(CassandraContainer container) {
+        LOGGER.info("Getting replacement requirement for task: {}", container.getId());
+        try {
+            return new OfferRequirement(
+                    clearTaskIds(container.getTaskInfos()),
+                    clearExecutorId(container.getExecutorInfo()),
+                    null,
+                    null);
+        } catch (InvalidRequirementException e) {
+            LOGGER.error("Failed to construct OfferRequirement with Exception: ", e);
+            return null;
+        }
     }
 
     public OfferRequirement getUpdateOfferRequirement(Protos.TaskInfo taskInfo) {
         LOGGER.info("Getting updated requirement for task: {}", taskInfo.getTaskId().getValue());
         return getExistingOfferRequirement(taskInfo);
     }
 
+    private List<Protos.TaskInfo> clearTaskIds(Collection<Protos.TaskInfo> taskInfos) {
+        List<Protos.TaskInfo> outTaskInfos = new ArrayList<>();
+        for (Protos.TaskInfo restartTaskInfo : taskInfos) {
+            outTaskInfos.add(
+                    Protos.TaskInfo.newBuilder(restartTaskInfo)
+                            .setTaskId(Protos.TaskID.newBuilder().setValue(""))
```

Comment:
What does hasXxx do for required fields after a clear? You can't build a
protobuf if it has an unset required field. I'm pretty sure hasXxx will
always return true, even after clear. Or maybe we won't be able to build
the protobuf.

hasXxx is cleaner but doesn't work for OfferEvaluator stuff. Empty vs
filled out has meaning. hasXxx doesn't tell us what we want to know,
because we can't build a protobuf that doesn't have a taskId or executorId
because they are required. I'm just unclear on the protobuf function
behavior.

Never set -> hasXxx is false
Set to empty -> hasXxx is true, but we don't check this anyway
clear() -> hasXxx is ???.

The approach used here is unambiguous.

On Mon, Jul 11, 2016 at 7:03 PM Mohit Soni <notifications@github.com> wrote:

> In
> cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/offer/PersistentOfferRequirementProvider.java
> <https://github.com/mesosphere/dcos-cassandra-service/pull/173#discussion_r70367505>
> :
>
> >      }
> >
> >      public OfferRequirement getUpdateOfferRequirement(Protos.TaskInfo taskInfo) {
> >          LOGGER.info("Getting updated requirement for task: {}", taskInfo.getTaskId().getValue());
> >          return getExistingOfferRequirement(taskInfo);
> >      }
> >
> > +    private List<Protos.TaskInfo> clearTaskIds(Collection<Protos.TaskInfo> taskInfos) {
> > +        List<Protos.TaskInfo> outTaskInfos = new ArrayList<>();
> > +        for (Protos.TaskInfo restartTaskInfo : taskInfos) {
> > +            outTaskInfos.add(
> > +                    Protos.TaskInfo.newBuilder(restartTaskInfo)
> > +                            .setTaskId(Protos.TaskID.newBuilder().setValue(""))
>
> Doing clearXXX gives us an advantage of using hasXxx which is less error
> prone and cleaner than doing a getXxx (which will always return a valid
> object in ProtoBuf world) and then checking whether it's empty.
> Clearing does guarantee that the underlying objects goes away, so
> effectively hasXxx will return false, post clearXxx operation. Also,
> doing an accidental getXxx (assuming it's a string), followed by an empty
> check will also be successful.
>
> Either ways, I'm okay.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mesosphere/dcos-cassandra-service/pull/173/files/8631b704b2f6dae59587246dcbd501967715c8d9#r70367505>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAinR8wsH3XYg9LAdtV2s8kV495c9y3bks5qUvXVgaJpZM4JJvNZ>
> .
>